### PR TITLE
feat: generate declaration files

### LINKED ISSUE #2: `.d.ts` generation

### PR DESCRIPTION
**Pull request by danielroe**
Reference: https://github.com/microsoft/TypeScript/wiki/Using-the-Compiler-API#getting-the-dts-from-a-javascript-file

Note that generating `.d.ts` files will ~~dramatically~~ increase the time required to run `mkdist` (from order of 50-100ms to ~~7-10s~~ 500ms).

**TODO**:
- [x] generates declarations for .vue files
- [x] generates declarations for .ts files
- [x] generates declarations for .js files
- [x] add cli option
- [ ] ~~use file cache rather than reading files from disk~~
- [ ] ~~enable by default~~

Closes #2

### CODE REVIEW

File: test/fixture/src/components/js.vue [hunk 1]
```diff
@@ -3,11 +3,14 @@
...
 <script>
-import { test } from '..'
+import test from '..'
+
+const str = 'test'
 
 export default {
-  data () {
-    test: test()
-  }
```
Comment (danielroe) [inline test/fixture/src/components/js.vue:11]: I've assumed this was a typo.

File: src/loaders/dts.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import type { CompilerOptions } from 'typescript'
+
+export async function getDeclaration (contents: string, path = '_contents.ts') {
+  try {
+    const { createCompilerHost, createProgram } = await import('typescript')
```
Comment (pi0) [inline src/loaders/dts.ts:5]: Maybe we can catch with a friendly error to install peer dependencies?

File: src/loaders/dts.ts [hunk 3]
```diff
@@ -27,6 +27,7 @@ export async function getDeclaration (contents: string, path = '_contents.ts') {
 
     return files[path.replace(/\.(ts|js)$/, '.d.ts')]
   } catch {
+    console.warn(`Could not generate declaration file for ${path}. Do you have \`typescript\` installed?`)
```
Comment (pi0) [inline src/loaders/dts.ts:30]: Exception catching is too board and we are hiding any other potential error that might be happen. We might move `Do you have` part directly after import chain and here include error as second arg

**Comment by danielroe** [comment]
I've managed to improve performance but it's still 10x longer to emit types than just transpile the files. I can imagine enabling it when using `siroc`, but perhaps we leave disabled by default when running `mkdist`?

File: src/loaders/js.ts [hunk 4]
```diff
@@ -21,6 +23,19 @@ export const jsLoader: Loader = async (input, { options }) => {
 
   let contents = await input.getContents()
 
+  const declaration: LoaderResult = []
+
+  if (options.emitTypes === true || (options.emitTypes === 'ts' && input.path.endsWith('.ts'))) {
```
Comment (pi0) [inline src/loaders/js.ts:28]: - Maybe we can use extname for extension 
- Shouldn't we always check extension name? (i didn't get the point of `true` vs `ts` but for vue loadAsset, we are using correct extension)

File: src/loaders/dts.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+const options: CompilerOptions = {
+  allowJs: true,
+  declaration: true,
+  incremental: true,
+  skipLibCheck: true,
+  emitDeclarationOnly: true
+}
+
+export async function getDeclaration (contents: string, path = '_contents.ts') {
```
Comment (pi0) [inline src/loaders/dts.ts:11]: Maybe we can move it to utils? :D

**Comment by pi0** [comment]
> I've managed to improve performance but it's still 10x longer to emit types than just transpile the files. I can imagine enabling it when using siroc, but perhaps we leave disabled by default when running mkdist?

@danielroe I was going to benchmark as well. Yes disabling by default makes sense. We may be able to reuse same instance of typescript compiler across files and do it off-thread using a worker

File: src/loaders/js.ts (hunk 4 above)
Comment (danielroe) [reply src/loaders/js.ts:28]: The option is only if we might want to disable emitting types for .js files - but happy to lose that option.

Comment (pi0) [reply src/loaders/js.ts:28]: We can reconsider approach for dts generation filter once we wanted to add feature. vue loader might send a hint about its original extension that is used in a filter

Comment (danielroe) [reply src/loaders/js.ts:28]: 👍 Let me know if you think we any of these changes need to be made in this PR.

**Comment by pi0** [comment]
Typescript compiler is anyway slow but performance should be fixed by 421b200306fad0887afc611b34fcfcb36c1dba2b. It think it would be nice we add support for `vue-dts-gen` too

**Pull request merged**